src/rust/engine/serverset/src/lib.rs

```diff
@@ -67,18 +67,62 @@ impl<T: Clone + Send + Sync + 'static> Clone for Serverset<T> {
 #[derive(Clone, Copy, Debug)]
 #[must_use]
 pub struct HealthReportToken {
-  index: usize,
+  server_index: usize,
 }
 
-struct Inner<T> {
-  connections: Vec<Backend<T>>,
+struct Connection<T: Clone> {
+  connection: T,
+  server_index: usize,
+}
 
+struct Inner<T: Clone> {
+  // Order is immutable through the lifetime of a Serverset, but contents of the Backend may change.
+  servers: Vec<Backend>,
   // Only visible for testing
-  pub(crate) next: usize,
+  pub(crate) next_server: usize,
+
+  // Points into the servers list; may change over time.
```

Comment:
And vice-versa: the length of this vec might change.

Reply:
Done